Open Problem: Mutable Data (Naming, Real-Time, Guarantees) (#6)
ipfs/research: OPEN_PROBLEMS/MUTABLE_DATA.md

Creating a sound solution to achieve Mutable Data for the dWeb requires considering the three parts of the problem:

- **Update Propagation** - As noted, diffent applications will have different requirementsm, these can go from real-time delivery, to delivery at least/most once, confirmation of delivery, conflict resolution, etc.
- **Update Authentication** - How to trust the parties that are propagating the updates, how to trust that they are not hiding any important update
- **Interface/API** - This is where things that shape up to feel like a database start to show. From convinence features to powerful APIs that enable a user to query, update and search over multiple records

All of these problems should be solved to work on a Distributed Network scenario, this means: disconnected, offline first and over unreliable networks.

## State of the Art

> This survey on the State of the Art is not by any means complete, however, it should provide a good entry point to learn what are the existing work. If you have something that is fundamentally missing, please consider submitting a PR to augment this survey. 

### Within the IPFS Ecosystem
> Existing attempts and strategies

> ##### IPNS

Author: @hugomrdias @aschmahmann where can I find your latest slides, talks and notes on IPNS with multiple routers?

Author: - Talk https://hackmd.io/@hugodias/ipns#/

Author: Also include:
- [Improve IPNS spec]https://github.com/ipfs/specs/issues/205)
- [IPNS-over-PubSub as an Independent Transport](https://github.com/ipfs/specs/pull/218)

Reviewer: There's also this issue: ipfs/specs#198

# Mutable Data (Naming, Real-Time, Guarantees)

## Short Description
> In one sentence or paragraph.

> How to enable a multitude of patterns of interaction between users and users, users and machines and machines and machines. In another words, what primitives to create to dynamic data apps and what guarantees can these offer.

Reviewer 2: ```suggestion
How to enable a multitude of patterns of interaction between users and users, users and machines and machines and machines. In other words, which primitives are necessary to create dynamic data apps and what guarantees can they offer?
```
The second sentence is not clear. Was this the intended meaning? It still doesn't read like a problem description.

## Long Description

> Mutable Data in a Permanent Datastore? WAT?! If this is your reaction, don't despair, we've been there too. However, the good news is that achieving mutability over immutable datastores is not a new idea, in fact, if you are checking this document through Git/GitHub, you are experiencing just that.

Reviewer 2: ```suggestion
Mutable Data in a Permanent Datastore? WAT?! If this is your reaction, don't despair, we've been there too. However, the good news is that achieving mutability over immutable datastores is not a new idea; in fact, if you are checking this document through Git/GitHub, you are experiencing just that.
```

The simple trick is to use pointers. While each data struct will have its own reference (in the case of IPFS, they will have a CID), mutability is achieved by updating the pointer to point to a new version of that data structure, just like when you update your master branch (pointer) to your latest commit (immutable reference) on your Git projects.

> Having Mutable Data is essential to create a whole new set of applications for the Web 3.0. These applications will have different types of requirements, from millions of records being update every second, to millions of users interacting and mutating a data source.

Reviewer 2: ```suggestion
Having Mutable Data is essential to create a whole new set of applications for the Web 3.0. These applications will have different types of requirements, from millions of records being updated every second to millions of users interacting and mutating a data source.
```

Creating a sound solution to achieve Mutable Data for the dWeb requires considering the three parts of the problem:

> - **Update Propagation** - As noted, diffent applications will have different requirementsm, these can go from real-time delivery, to delivery at least/most once, confirmation of delivery, conflict resolution, etc.

Reviewer 2: ```suggestion
- **Update Propagation** - As noted, different applications will have different requirements, e.g. real-time delivery, delivery at least/most once, confirmation of delivery, conflict resolution, etc.
```

> - **Update Authentication** - How to trust the parties that are propagating the updates, how to trust that they are not hiding any important update

Reviewer 2: To be clear: do we mean trust that their updates are correct **and** trust that they're not withholding updates?

Author: Exactly.

> - **Interface/API** - This is where things that shape up to feel like a database start to show. From convinence features to powerful APIs that enable a user to query, update and search over multiple records

Reviewer 2: ```suggestion
- **Interface/API** - Supporting a database-link experience, including APIs that enable a user to query, update and search over multiple records
```
It wasn't reading like a part of the problem. Not sure this is any better.

Author: What's a `database-link`?

> All of these problems should be solved to work on a Distributed Network scenario, this means: disconnected, offline first and over unreliable networks.

Reviewer 2: ```suggestion
All of these problems should be solved in a manner that is tolerant of unreliable connections and offline operation.
```

That is definitely not the meaning of a distributed network. Does the suggestion communicate all you want? An alternative may be needed.

- https://github.com/libp2p/specs/blob/master/pubsub/README.md
- https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md

##### peer-base's CRDTs implementations

CRDTs are an extraordinary building block for data mutability without requiring central mediation. The field of CRDT is vast and there has been many developments. Below you can find some of the implementations done to work with IPFS.

- https://github.com/ipfs-shipyard/js-delta-crdts
- https://github.com/ipfs-shipyard/peer-crdt 
- https://github.com/ipfs-shipyard/peer-crdt-ipfs
- https://orbitdb.org/

##### Textile's data wallet and caffee 

Textile offers a set of tools for build decentralized applications that also work on mobile platforms. Their documentation is pretty amazing and keeps evolving, to learn more follow the links below:

- https://docs.textile.io/concepts/the-wallet
- https://docs.textile.io/concepts/cafes
- [Tutorial on how to use Textile form IPFS Camp](https://github.com/ipfs/camp/tree/master/CORE_AND_ELECTIVE_COURSES/ELECTIVE_COURSE_D)

### Within the broad Research Ecosystem
> How do people try to solve this problem?

> There are multiple decades of Research and Projects to solve this Open Problem. Don't feel disencourged by the sheer amount of articles, it is really high valuable work!

Reviewer 2: ```suggestion
There are multiple decades of Research and Projects to solve this Open Problem. Don't feel discouraged by the sheer amount of articles, it is really high valuable work!
```

  - [Snapdoc: Authenticated snapshots with history privacy in peer-to-peer collaborative editing](https://martin.kleppmann.com/papers/snapdoc-pets19.pdf)
  - [OpSets: Sequential Specifications for Replicated Datatypes](https://arxiv.org/abs/1805.04263)
- Access Control
  - [Access Control for Weakly Consistent Data Stores](http://www.complang.tuwien.ac.at/kps2015/proceedings/KPS_2015_submission_25.pdf)
  - [ACGreGate: A Framework for Practical Access Control for Applications using Weakly Consistent Databases](https://arxiv.org/abs/1801.07005)
  - [TRVE Data: Placing a bit less trust in the cloud](https://www.cl.cam.ac.uk/research/dtg/trve/)
  - [LSEQ: an Adaptive Structure for Sequences in Distributed Collaborative Editing](https://hal.archives-ouvertes.fr/hal-00921633/document)
- Applications & Libraries
  - [A simple approach to building a real-time collaborative text editor](http://digitalfreepen.com/2017/10/06/simple-real-time-collaborative-text-editor.html)
  - http://y-js.org/
  - http://swarmdb.net/
  - http://gun.js.org/
  - http://scuttlebot.io/
  - https://github.com/mafintosh/hyperlog
  - https://github.com/jboner/akka-crdt

> ### Known shortcommins of existing solutions

Reviewer 2: ```suggestion
### Known shortcomings of existing solutions
```

> What are the limitations on those solutions?

> Given the sheer amount of solutions and experiments to solve the multiple challenges of this Open Problem, one common shortcomming is actually a qualitative and quantitative assessement of what is possible today and how these stack up to existing centralized versions and if they are capable or not to take on the load to fulfil the needs of the multiple use cases.

Reviewer 2: ```suggestion
Given the sheer amount of solutions and experiments to solve the multiple challenges of this Open Problem, one common shortcoming is actually a qualitative and quantitative assessment of what is possible today and how these stack up to existing centralized versions and if they are capable or not to take on the load to fulfil the needs of the multiple use cases.
```
This feels weirdly vague. Aren't we aware of the specific shortcomings of e.g. the existing IPNS implementation? This (not making it clear what's currently missing) has been a recurring issue in the specification of open problems and RFPs. Fine if you tell me there's no point in discussing IPNS faults if we don't have a complete survey of the alternatives.

## Solving this Open Problem

### What is the impact

> Mutable Data is a fundamental building block for creating applications. Improving the existing solutions and creating new ones that will support a large number of users in a distributed context, will enable the entry of the large growing number of Internet users to be part of the dWeb.

Reviewer 2: ```suggestion
Mutable Data is a fundamental building block for creating applications. Improving the existing solutions and creating new ones that support a large user set in a distributed context will enable the growing number of Internet users to be part of the dWeb.
```

> ### What defines a complete solution?

Author: I'm not necessarily happy with this section. I feel it is not clear enough on how ambitious and challenging solving this Open Problem is.

@jsoares @yiannisbot Would you like to review and make some suggestions? I'm out of creative juice :D

Reviewer: TLDR: A complete solution for mutable data (in a p2p context) is pretty unreasonable given that we don't have a one-size-fits-all solution in the centralized world.

I suspect that the reason it's not clear is that there isn't (or has not yet been defined) a one-size fits all answer for what mutable structures should look like or how they should work. Some examples:

- What form of access control/capabilities even makes in a non-centralized context?
   - Consensus, eventual consistency, causal consistency, etc.
   - Should ACLs be changeable?
   - If cryptography is utilized for ACLs how is it upgraded?
- What assumptions can people work with?
   - This problem maps really closely with solving distributed identity, but most solutions assume pre-distributed identity... it's a bit of a contradiction
   - If cryptography is utilized which types are ok, which math is sufficient
- How do people find new mutable data?
   - This might seem unrelated, but given its tie into the identity problem as well as potential forking of mutable data streams it's pretty important

I suspect that step 1 in any "complete solution" is defining the tradeoffs and composability properties of solution categories. This is perhaps a combination of something like ipfs/notes#379 with a more precise version of the Zooko's triangle tradeoffs (it is notoriously underdefined/specified and could probably be more complete if explicitly correlated to the CAP theorem).

Author: Thank you for the notes @aschmahmann! :)

Author: Took your notes, restructured a bit and expanded the section more. Thank you!

@yiannisbot @jsoares your feedback is still very welcome. I'll go ahead and call this one done, PRs are always accepted :)

Reviewer 2: I have a bunch of language fixes but those can wait for @yiannisbot's comments. The version incorporating Adin's comments is already much more detailed but, as acceptance criteria for an RFP, it's still fuzzy. 

Saying that being the easy part, coming up with objective criteria probably requires some discussion of system parameters and requirements, which is not critical at this stage but should be done at the RFP drafting stage (see nice example [here](https://github.com/protocol/research-RFPs/blob/master/RFPs/rfp-6-SSLE.md#requirements--constraints)).

Author: > as acceptance criteria for an RFP, it's still fuzzy.
> ..
> which is not critical at this stage but should be done at the RFP drafting stage (see nice example here).

Exactly, this docs is not the RFP and more broad in nature
